Final Refactoring - Syw UID2-4159 token gen code refactoring user identity

sunnywu opened the pull request:
1. Read https://thetradedesk.atlassian.net/browse/UID2-4159
3. Refactor EncryptedTokenEncoder#encodeIntoIdentityResponse
4. Moved toJson methods into IdentityResponse from UIDOperatorVerticle class
5. [Use EncryptedTokenEncoder directly inside UIDOperatorService class instead of ITokenEncoder interfact so it's easier to jump to references and we haven't really created any other ITokenEncoder varantsi](https://github.com/IABTechLab/uid2-operator/pull/1081/commits/131d203972adf6948380fb1c559de812eb5bc028)
1. Removed the UserIdentity's refreshedAt field as that's redundant and currently set to the same value as the AdvertisingTokenInput/RefreshTokenInput#createdAt timestamp.
2. Renamed some `token` variable names to adTokenInput and refreshTokenInput to make it clearer
3. Removed privacyBits and establishedAt from UserIdentity/HashedDiidentity/RawUidIdentity. EstbalishedAt is kept inside FirstLevelHashIdentity as the time the first level hash was created..
5.  Also, privacyBits and establishedAt are generated from either a. in a brand new token generation call (by default should be 1 and the timestamp at the time of the call) b. or, during token refresh logic, these fields will be inherited from previous refresh token.
6. RefreshTokenRequest, AdvertisingTokenRequest and IdentityRequest will now require to store privacyBits and establishedAt itself as a compromise. 
7. generateIdentity will now take in a privacyBits param.
9. Created verifyFirstLevelHashIdentityAndEstablishedAt method for verifying FirstLevelHashIdentity and added more checks for FirstLevelHash and check the correctness of raw Uid/First level hash in various unit tests
10. Added PrivacyBitsTest/IdentityResponseTest/RawUidResponseTest classes and added some extra check for firstLevelHash in existing tests
11.  Replace `int` privacy bits type use to proper PrivacyBits class instance
12. Refactored UIDOperatorVerticleTest unit tests to standardise the advertisingTokenRequest/refreshTokenRequest verification(https://github.com/IABTechLab/uid2-operator/pull/1123/commits/2edcb05239c63ca64051a03627d6ae7b868a607b)

Changes from 2024-12-01
14. Renamed TokenUtils.java's functions to be blatently clear on the input value type
15. Fixed the old testIdentityMapForOptOutUser test that was using raw DII as a hash
16. Renamed all uses of IdentityString/IdentityHash variables to rawDii and hashedDii
17. Renamed IdentityType to DiiType
18.  In InputUtil, renamed Identity* member variable/methods to Dii* to indicate it's Dii
19. Change TokenUtils class getIdentityHash to getDiiHash and same for getDiiHashString
20. Renamed FirstLevelHashIdentity/HashedDiiIdentity/RawUidIdentity to FirstLevelHash/HashedDii/RawUid
21. Renamed RawUidResponse to IdentityMapResponseItem, MapRequest to IdentityMapRequestItem
22. Renamed IUIDOperatorService.mapHashedDiiIdentity to mapHashedDii
23. Renamed IdentityRequest/IdentityResponse to TokenGenerateRequest/Response
24. Renamed RefreshTokenRequest/RefreshResponse to     TokenRefreshRequest/TokenRefreshResponse
25. Renamed AdvertisingTOkenInput/RefreshTokenInput to AdvertisingTokenRequest/RefreshTOkenRequest

sunnywu committed:
Renamed IdentityToken class and variables to Identity as it's a set of variables which don't just include tokens

sunnywu committed:
Renamed Identity.LogoutToken to InvalidIdentity and added some comments for clarity

sunnywu committed:
Renamed Identity.LogoutToken to InvalidIdentity and added some comments for clarity

sunnywu committed:
renamed getAdvertisingId to MappedIdentity

sunnywu committed:
Split UserIdentity into 3 distinct classes - HashedDiiIdentity, FirstLevelHashIdentity and RawUidIdentity as UserIdentity is a very vague term and one needs to read function calls to really understand whether the UserIdentity class has hashed DII, first level hash or raw Uid.

sunnywu committed:
Added brief comment to Identity/MappedIdentity

sunnywu committed:
Renamed MappedIdentity to MappedIdentityResult

sunnywu committed:
1. Rename AdvertisingToken class to AdvertisingTokenInput
2. Rename RefreshToken class to RefreshTokenInput
3. Rename createAdvertisingToken/createRefreshToken methods to createAdvertisingTokenInput/createRefreshTokenInput

sunnywu committed:
Renamed ITokenEncoder.encode to ITokenEncoder.encodeIntoIdentityResponse

sunnywu committed:
Renamed Identity.isEmptyToken method to Identity.isNotValid

sunnywu committed:
Renamed PublisherIdentity to SourcePublisher

sunnywu committed:
Refactored UserIdentity as a base class which is inherited by HashedDii/FirstLevelHash/RawUidIdentity sub-classes

sunnywu committed:
Renamed Identity to IdentityResponse

sunnywu committed:
Fixed using raw identity in testGenerateTokenForOptOutUser

sunnywu committed:
Renamed RefreshResponse#getIdentity to RefreshResponse#getIdentityResponse

sunnywu committed:
1. Refactor EncryptedTokenEncoder#encodeIntoIdentityResponse
2. Moved toJson methods into IdentityResponse from UIDOperatorVerticle class

sunnywu committed:
Use EncryptedTokenEncoder directly inside UIDOperatorService class instead of ITokenEncoder interfact so it's easier to jump to references and we haven't really created any other ITokenEncoder variants

sunnywu committed:
Renamed MappedIdentity to RawUid2Result and  UIDOperatorService.getAdvertisingId to generateRawUid

sunnywu committed:
1. Fixed RawUid2Result to RawUidResult
2. Renamed AdvertisingId to RawUid in most operator codes (except optout store related)

sunnywu committed:
Merge branch 'syw-UID2-4159-token-gen-code-renaming' into syw-UID2-4159-token-gen-code-refactoring

sunnywu committed:
Renamed IdentityResponse.optOutIdentityResponse to OptOutIdentityResponse

sunnywu committed:
Renamed IdentityResponse.optoutIdentityResponse to OptOutIdentityResponse

sunnywu committed:
Merge branch 'syw-UID2-4159-token-gen-code-renaming' into syw-UID2-4159-token-gen-code-refactoring

sunnywu committed:
Fixed UserIdentity member variables as final and added constructor

sunnywu committed:
Fixed UserIdentity member variables as final and added constructor

sunnywu committed:
1. Addressed code review feedbacks
2. Renamed RawUidResult to RawUidResponse
3. rename some id variable to rawUid to make it clearer
4. more comments

sunnywu committed:
1. Addressed code review feedbacks
2. Renamed RawUidResult to RawUidResponse
3. rename some id variable to rawUid to make it clearer
4. more comments

sunnywu committed:
Made UserIdentity abstract first

sunnywu committed:
Made UserIdentity abstract first

sunnywu committed:
Fixed a few id variable to rawUid/firstLevelHash to make it clear. Fixed the UserIdentity/FirstLevelHashIdentity class uses to HashedDiiIdentity instead in IdentityMapBenchmark and TokenEndecBenchmark

sunnywu committed:
Fixed a few id variable to rawUid/firstLevelHash to make it clear. Fixed the UserIdentity/FirstLevelHashIdentity class uses to HashedDiiIdentity instead in IdentityMapBenchmark and TokenEndecBenchmark

sunnywu committed:
Merge remote-tracking branch 'origin/syw-UID2-4159-token-gen-code-renaming' into syw-UID2-4159-token-gen-code-refactoring

sunnywu committed:
1. Refactored EncryptedTokenEncoder to rename encode/encodev2/encodev3
   methods to encodeIntoAdvertisingToken or encodeIntoRefreshToken to
   make it clear what we are encoding into

sunnywu committed:
Merge remote-tracking branch 'origin/syw-UID2-4159-token-gen-code-renaming' into syw-UID2-4159-token-gen-code-refactoring

sunnywu committed:
1. Removed the UserIdentity's refreshedAt field as that's redundant and
   currently set to the same value as the AdvertisingTokenInput/RefreshTokenInput#createdAt
   timestamp.
2. Renamed some `token` variable names to adTokenInput and
   refreshTokenInput to make it clearer

sunnywu committed:
1. Removed privacyBits and establishedAt from UserIdentity/HashedDiiIdentity/RawUidIdentity but only kept inside FirstLevelHashIdentity - these fields are only relevant when a first level hash is generated for token generation and then passed along into ad token generation logic.
2. As a result,  privacyBits and establishedAt are generated from either
   a. in a brand new token generation call (by default should be 1 and
   the timestamp at the time of the call)
   b. or, during token refresh logic, these fields will be inherited
   from previous refresh token.
3. Updated a lot of codes as a result (more refinement required such as
   not hardcoding the "1, Instant.now()" in a lot of places)

sunnywu committed:
added comments

sunnywu committed:
fixed variable name

sunnywu committed:
fixed variable name

sunnywu committed:
Merge remote-tracking branch 'origin/syw-UID2-4159-token-gen-code-renaming' into syw-UID2-4159-token-gen-code-refactoring

sunnywu committed:
Merge remote-tracking branch 'origin/syw-UID2-4159-token-gen-code-refactoring' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Moved privacyBits out of FirstLevelHashIdentity and treat it as a separate input

sunnywu committed:
Added a new IdentityRequest constructor to set default values for privacyBits and establishedAt fields

sunnywu committed:
Update src/main/java/com/uid2/operator/model/IdentityRequest.java

Co-authored-by: mcollins-ttd <118872455+mcollins-ttd@users.noreply.github.com>

sunnywu committed:
Replace  privacy bits type to PrivacyBits class instance

sunnywu committed:
Replace  privacy bits type to PrivacyBits class instance

sunnywu committed:
Replace  privacy bits type to PrivacyBits class instance

sunnywu committed:
missing a newline

sunnywu committed:
1. Added PrivacyBitsTest/IdentityResponseTest/RawUidResponseTest classes
2. added some extra check for firstLevelHash in existing tests

sunnywu committed:
Created verifyFirstLevelHashIdentityAndEstablishedAt method for verifying FirstLevelHashIdentity and added more checks for FirstLevel and check the correctness of raw Uid/First level hash in various unit tests

sunnywu committed:
Created verifyFirstLevelHashIdentityAndEstablishedAt method for verifying FirstLevelHashIdentity and added more checks for FirstLevel and check the correctness of raw Uid/First level hash in various unit tests

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu commented on src/main/java/com/uid2/operator/model/AdvertisingToken.java:
```diff
@@ -1,19 +0,0 @@
-import java.time.Instant;
-import com.uid2.shared.model.TokenVersion;
-
-public class AdvertisingToken extends VersionedToken {
```
renamed to AdvertisingTokenInput

sunnywu commented on src/main/java/com/uid2/operator/model/IdentityResponse.java:
```diff
@@ -0,0 +1,77 @@
+
+// this defines all the fields for the response of the /token/generate and /client/generate endpoints before they are
+// jsonified
+public class IdentityResponse {
```
was IdentityTokens

sunnywu replied:
renamed to `TokenGenerateResponse`

sunnywu commented on src/main/java/com/uid2/operator/model/MappedIdentity.java:
```diff
@@ -1,16 +0,0 @@
-package com.uid2.operator.model;
-
-public class MappedIdentity {
```
renamed to RawUidResponse

sunnywu commented on src/main/java/com/uid2/operator/model/RawUidResponse.java:
```diff
@@ -0,0 +1,19 @@
+package com.uid2.operator.model;
+
+// Contains the computed raw UID and its bucket ID from identity/map request
+public class RawUidResponse {
```
renamed from MappedIdentity

gmsdelmundo replied:
Just checking again - I think `MapRequest` is the request object for `identity/map` right? We don't want to have the same name here? like `IdentityRequest`/`IdentityResponse` or `RefreshTokenInput`/`RefreshResponse`

gmsdelmundo replied:
I guess this applies too to the other `TokenInput` classes

sunnywu replied:
hmm the identity/map function has a hashed dii input and returns a raw Uid (and bucket id). I can call `RawUidResponse` to `MapResponse` but i think `RawUidResponse` is clearer so changing to `MapResponse` for the sake matching up is not necessary making it clearer and therefore i would hesitate to make this change. 

For `AdvertisingTokenInput/RefreshTokenInput`, i have renamed to `AdvertisingTokenRequest/RefreshTokenRequest`

sunnywu replied:
Renamed to IdentityMapResponseItem and also MapRequest to IdentityMapRequestItem
I renamed IUIDOperatorService.mapHashedDiiIdentity to mapHashedDii as well

sunnywu commented on src/main/java/com/uid2/operator/model/RefreshToken.java:
```diff
@@ -1,18 +0,0 @@
-import java.time.Instant;
-import com.uid2.shared.model.TokenVersion;
-
-public class RefreshToken extends VersionedToken {
```
Renamed to RefreshTokenInput

sunnywu commented on src/main/java/com/uid2/operator/model/RefreshTokenInput.java:
```diff
@@ -0,0 +1,25 @@
+import com.uid2.operator.util.PrivacyBits;
+import com.uid2.shared.model.TokenVersion;
+
+public class RefreshTokenInput extends VersionedToken {
```
Renamed from RefreshToken

gmsdelmundo replied:
What about `RefreshTokenRequest` instead? So it matches the other request-type objects

sunnywu replied:
renamed to ```TokenRefreshRequest``` now

sunnywu commented on src/main/java/com/uid2/operator/model/UserIdentity.java:
```diff
@@ -1,30 +0,0 @@
-import java.util.Arrays;
-import java.util.Objects;
-
-public class UserIdentity {
```
Become an abstract base class and moved into the userIdentity sub-package

sunnywu committed:
Checking privacy bits in more tests in UIDOperatorVerticleTest

Commit:
[CI Pipeline] Released Snapshot version: 5.41.1-alpha-116-SNAPSHOT

sunnywu committed:
revered version in pom.xml

sunnywu commented on src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:
```diff
@@ -936,13 +939,12 @@ private void handleTokenGenerateV1(RoutingContext rc) {
+                final IdentityResponse t = this.idService.generateIdentity(
                         new IdentityRequest(
-                                new PublisherIdentity(siteId, 0, 0),
-                                input.toUserIdentity(this.identityScope, 1, Instant.now()),
```
`1, Instant.now()` became the default param for IdentityRequest constructor now

sunnywu commented on src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:
```diff
@@ -989,13 +991,13 @@ private void handleTokenGenerateV2(RoutingContext rc) {
+                final IdentityResponse t = this.idService.generateIdentity(
                         new IdentityRequest(
-                                new PublisherIdentity(siteId, 0, 0),
-                                input.toUserIdentity(this.identityScope, 1, Instant.now()),
```
1, Instant.now() became the default param for IdentityRequest constructor now

sunnywu committed:
added some comments and renamed refreshIdentity method param to input instead of 	oken

sunnywu commented on src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:
```diff
@@ -1235,11 +1237,11 @@ private void handleIdentityMapV1(RoutingContext rc) {
         }
         try {
             final Instant now = Instant.now();
-            final MappedIdentity mappedIdentity = this.idService.map(input.toUserIdentity(this.identityScope, 0, now), now);
```
identity/map call doesn't use the 0 privacy bits and established at timestamp

sunnywu commented on src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:
```diff
@@ -48,13 +55,13 @@ public class UIDOperatorServiceTest {
     ExtendedUIDOperatorService euidService;
     Instant now;
-
+    RotatingSaltProvider saltProvider;
```
moved here so we could use it to verify generated first level hash/uid result

sunnywu committed:
fixed unit tests checking establishedAt between advertisingTokenInput & refreshTokenInput's

sunnywu committed:
Refactored unit test codes to standardise the advertisingTokenInput/refreshTokenInput verification

sunnywu committed:
Refactored unit test codes to standardise the advertisingTokenInput/refreshTokenInput verification

asloobq commented on src/main/java/com/uid2/operator/service/EncryptedTokenEncoder.java:
```diff
@@ -117,17 +122,19 @@ private RefreshToken decodeRefreshTokenV2(Buffer b) {
-                new PublisherIdentity(siteId, 0, 0),
-                new UserIdentity(IdentityScope.UID2, IdentityType.Email, identity, privacyBits, Instant.ofEpochMilli(establishedMillis), null));
+                new SourcePublisher(siteId, 0, 0),
+                new FirstLevelHashIdentity(IdentityScope.UID2, IdentityType.Email, identity,
```
why do we assum that the scope is UID2 and not EUID ?

asloobq replied:
got it. We can clean up the old code in the near future.

asloobq reviewed (commented):
This really makes it easy to understand the core raw UID and token generation. Thanks for doing this.

sunnywu committed:
Code review feedback

sunnywu commented on src/main/java/com/uid2/operator/service/IUIDOperatorService.java:
```diff
@@ -11,22 +12,22 @@
+    RefreshResponse refreshIdentity(RefreshTokenInput input);
 
-    MappedIdentity mapIdentity(MapRequest request);
+    RawUidResponse mapIdentity(MapRequest request);
```
changed to `mapHashedDiiIdentity` to stay consistent with `generateIdentity`/`refreshIdentity`

sunnywu committed:
renamed IUIDOperatorService#mapIdentity to mapHashedDiiIdentity

sunnywu commented on src/main/java/com/uid2/operator/service/UIDOperatorService.java:
```diff
@@ -221,69 +226,79 @@ public Duration getIdentityExpiryDuration() {
     }
 
-    private UserIdentity getFirstLevelHashIdentity(IdentityScope identityScope, IdentityType identityType, byte[] identityHash, Instant asOf) {
+    private FirstLevelHashIdentity getFirstLevelHashIdentity(IdentityScope identityScope, IdentityType identityType, byte[] identityHash, Instant asOf) {
```
a common functions used by opt out identity creation like UIDOperatorService#testOptOutIdentityForEmail
https://github.com/IABTechLab/uid2-operator/blob/c6586a55e3f4226470a3b25269dbbe84c819374c/src/main/java/com/uid2/operator/service/UIDOperatorService.java#L70

sunnywu commented on src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:
```diff
@@ -989,13 +991,13 @@ private void handleTokenGenerateV2(RoutingContext rc) {
                 }
 
-                final IdentityTokens t = this.idService.generateIdentity(
+                final IdentityResponse t = this.idService.generateIdentity(
```
fixed

sunnywu commented on src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:
```diff
@@ -1045,14 +1047,14 @@ else if (!input.isValid()) {
         try {
             siteId = AuthMiddleware.getAuthClient(rc).getSiteId();
-            final IdentityTokens t = this.idService.generateIdentity(
+            final IdentityResponse t = this.idService.generateIdentity(
```
renamed to response

sunnywu committed:
renamed  to

gmsdelmundo commented on src/main/java/com/uid2/operator/model/RefreshResponse.java:
```diff
@@ -4,37 +4,37 @@
-    public static RefreshResponse Expired = new RefreshResponse(Status.Expired, IdentityTokens.LogoutToken);
-    public static RefreshResponse Deprecated = new RefreshResponse(Status.Deprecated, IdentityTokens.LogoutToken);
-    public static RefreshResponse NoActiveKey = new RefreshResponse(Status.NoActiveKey, IdentityTokens.LogoutToken);
+    public static RefreshResponse Invalid = new RefreshResponse(Status.Invalid, IdentityResponse.OptOutIdentityResponse);
```
I don't think these ever get mutated right? I think we can set these to `public static final`

sunnywu replied:
done

gmsdelmundo commented on src/main/java/com/uid2/operator/store/CloudSyncOptOutStore.java:
```diff
@@ -86,15 +86,15 @@ public long getOptOutTimestampByAdId(String adId) {
         this.webClient.get(remoteApiPort, remoteApiHost, remoteApiPath).
-            addQueryParam("identity_hash", EncodingUtils.toBase64String(firstLevelHashIdentity.id))
-            .addQueryParam("advertising_id", EncodingUtils.toBase64String(advertisingId))
+            addQueryParam("identity_hash", EncodingUtils.toBase64String(firstLevelHashIdentity.firstLevelHash))
```
Minor point - can we move the `.` from the previous line to this one, i.e. `.addQueryParam` instead of just `addQueryParam`

sunnywu replied:
Fixed

sunnywu committed:
Code review feedback and renamed AdvertisingTOkenInput/RefreshTokenInput to AdvertisingTokenRequest/RefreshTOkenRequest

asloobq reviewed (dismissed):
LGTM.

sunnywu commented on src/main/java/com/uid2/operator/model/SourcePublisher.java:
```diff
@@ -1,11 +1,12 @@
     public final long publisherId;
 
-    public PublisherIdentity(int siteId, int clientKeyId, long publisherId) {
+    public SourcePublisher(int siteId, int clientKeyId, long publisherId) {
```
i have added this comment in the class:

> 
>     // these 2 values are added into adverting/UID token and refresh token payload but
>     // are not really used for any real purposes currently so sometimes are set to 0
>     // see the constructor below

sunnywu committed:
Code review feedback,
Mainly renaming RefreshTokenRequest/RefreshResponse to
TokenRefreshRequest/TokenRefreshResponse

sunnywu commented on src/main/java/com/uid2/operator/IdentityConst.java:
```diff
@@ -3,18 +3,22 @@
     public static final byte[] ValidateIdentityForEmailHash = EncodingUtils.getSha256Bytes(IdentityConst.ValidateIdentityForEmail);
     public static final byte[] ValidateIdentityForPhoneHash = EncodingUtils.getSha256Bytes(IdentityConst.ValidateIdentityForPhone);
+
+    // DIIs to use when you want to generate a optout response in token generation or identity map
```
fixed

sunnywu committed:
Renamed IdentityRequest/IdentityResponse to TokenGenerateRequest/Response

sunnywu committed:
- Renamed RawUidResponse to IdentityMapResponseItem, MapRequest to IdentityMapRequestItem
- Renamed IUIDOperatorService.mapHashedDiiIdentity to mapHashedDii

sunnywu commented on src/main/java/com/uid2/operator/model/MapRequest.java:
```diff
@@ -1,18 +1,19 @@
+
 import java.time.Instant;
 
 public final class MapRequest {
```
renamed to ```IdentityMapRequestItem```

sunnywu committed:
1. Renamed FirstLevelHashIdentity/HashedDiiIdentity/RawUidIdentity to FirstLevelHash/HashedDii/RawUid
2. Removed the abstract class UserIdentity
3. fixed a unit test as removing UserIdentity broke it

sunnywu committed:
1. Renamed IdentityType to DiiType
2. In InputUtil, renamed Identity* member variable/methods to Dii* to indicate it's Dii
3. Change TokenUtils class getIdentityHash to getDiiHash and same for getDiiHashString

sunnywu committed:
1. Renamed TokenUtils.java's functions to be blatently clear on the input value type
2. Fixed the old testIdentityMapForOptOutUser test that was using raw DII as a hash
3. Renamed all uses of IdentityString/IdentityHash variables to rawDii and hashedDii

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Merge branch 'syw-UID2-4159-token-gen-code-refactoring-UserIdentity' of https://github.com/IABTechLab/uid2-operator into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

sunnywu committed:
Merge remote-tracking branch 'origin/main' into syw-UID2-4159-token-gen-code-refactoring-UserIdentity

Commit:
[CI Pipeline] Released Snapshot version: 5.56.43-alpha-201-SNAPSHOT
